index.d.ts

```diff
@@ -1,288 +1,591 @@
+  canFilter: boolean
+  setFilter: (
+    updater: ((filterValue: FilterValue) => FilterValue) | FilterValue
+  ) => void
+  filterValue: FilterValue
+  preFilteredRows: Row<D>[]
+}
 
-  export interface UseGroupByState {
-    groupBy: string[]
-    isAggregated?: boolean
-  }
+export type FilterProps<D extends object> = HeaderProps<D>
+export type FilterValue = any
+export type Filters<D extends object> = Record<IdType<D>, FilterValue>
+
+export type DefaultFilterTypes =
+  | 'text'
+  | 'exactText'
+  | 'exactTextCase'
+  | 'includes'
+  | 'includesAll'
+  | 'exact'
+  | 'equals'
+  | 'between'
+
+export interface FilterType<D extends object> {
+  (
+    rows: Row<D>[],
+    columnId: IdType<D>,
+    filterValue: FilterValue,
+    column: ColumnInstance<D>
+  ): Row<D>[]
+  autoRemove?: (filterValue: FilterValue) => boolean
+}
+/* #endregion */
 
-  export interface UseExpandedValues {
-    toggleExpanded?: () => any
-  }
+/* #region useGroupBy */
+export function useGroupBy<D extends object = {}>(hooks: Hooks<D>): void
+export namespace useGroupBy {
+  const pluginName = 'useGroupBy'
+}
 
-  export interface UseSortbyOptions {
-    sortByFn?: SortByFn
-    manualSorting?: boolean
-    disableSorting?: boolean
-    defaultSortDesc?: boolean
-    disableMultiSort?: boolean
-  }
+export type UseGroupByOptions<D extends object> = Partial<{
+  manualGroupBy: boolean
+  disableGrouping: boolean
+  aggregations: Record<string, AggregatorFn<D>>
+  groupByFn: (rows: Row<D>[], columnId: IdType<D>) => Record<string, Row<D>>
+}>
+
+export interface UseGroupByHooks<D extends object> {
+  getGroupByToggleProps: ((
+    header: HeaderGroup<D>,
+    instance: TableInstance<D>
+  ) => object)[]
+}
 
-  export interface UseSortbyState<D> {
-    sortBy?: SortingRules<D>
-  }
+export interface UseGroupByState<D extends object> {
+  groupBy: IdType<D>[]
+}
 
-  export interface TableInstance<D = {}> extends TableOptions<D> {
-    hooks: Hooks
-    rows: Row<D>[]
-    columns: EnhancedColumn<D>[]
-    headerGroups: HeaderGroup<D>[]
-    headers: HeaderGroup<D>[]
-    getTableProps: (userProps?: any) => any
-    getRowProps: (userProps?: any) => any
-    prepareRow: (row: Row<D>) => any
-  }
+export type UseGroupByColumnOptions<D extends object> = Partial<{
+  aggregate: Aggregator<D> | Aggregator<D>[]
+  Aggregated: Renderer<CellProps<D>>
+  disableGrouping: boolean
+  groupByBoundary: boolean
+}>
+
+export interface UseGroupByInstanceProps<D extends object> {
+  rows: Row<D>[]
+  preGroupedRows: Row<D>[]
+  toggleGroupBy: (columnId: IdType<D>, toggle: boolean) => void
+}
 
-  export interface TableOptions<D = {}> {
-    data: D[]
-    columns: HeaderColumn<D>[]
-    state: State<D>
-    debug?: boolean
-    loading: boolean
-    defaultColumn?: Partial<Column<D>>
-  }
+export interface UseGroupByColumnProps<D extends object> {
+  canGroupBy: boolean
+  isGrouped: boolean
+  groupedIndex: number
+  toggleGroupBy: () => void
+}
```

Comment:
I think that `getGroupByToggleProps: (props?: object) => object` also wants to be on the UseGroupByColumnProps interface since it appears to be in both places.

Reply:
I'm unable to find anything supporting that this is on the column instance

Reply:
There is usage like https://github.com/tannerlinsley/react-table/blob/master/examples/kitchen-sink/src/App.js#L326 in several of the examples.

Reply:
The example is out of date. The actual code doesn't add the prop to a column. It only adds it to the header.

Reply:
As for the prepareRow returned a boolean part, I think they were just using the fact that it doesn't have a return to automatically fallback to the else case of the OR statement. I do this quite often even if it doesn't return a boolean.

Reply:
that's as may be, but the function is actually on the column object (which I'll admit might actually be a header), and the code works.

I do admit that there's a bit of a conflict here between writing the interface to match what the documented interface says should be valid, vs deducing the interface based on how it's used.

Reply:
It's hard to argue the code when it only specifies it on the header. 
https://github.com/tannerlinsley/react-table/blob/f70d38e061ca6bae328e141d4c3c8efa1e9c86cd/src/plugin-hooks/useGroupBy.js#L129

Maybe @tannerlinsley can shed some light into this issue?

Reply:
FYI I think that the issue might stem from the fact that in the main table code, the terms header and column are used somewhat interchangeably.  useGroupBy adds `getGroupByToggleProps` to an object that it calls `header` that it pulls from `flatHeaders`.  In other parts of the code `flatHeaders` is iterated as a set of `columns`.  It makes keeping track of these things in the code tricky and a bit misleading.

Reply:
Agreed, I found that headers are essentially forked from columns. I'm not sure if the distinction needs to be clear or not.

Reply:
yeah, I'm not clear on that either :(